<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Meta-analysis of multiple populations along with GxE</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>



<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Meta-analysis of multiple populations along with GxE</h1>

<h2>Genome-wide association</h2>

<p>Let the dataset have \(n\) observations, \(f\) subpopulations, \(m\) markers and \(e\) environments. The method starts from the genome-wide association analysis in the \(j^{th}\) environment, following the alternative model \(y_j = \mu_j + Z\hat{\gamma}_i + \xi_j + \epsilon_j\). Where \(y\) is the vector corresponding to the response variable, \(\mu\) is the intercept, \(Z\) is a \(n\times f\) incidence matrix indicating the haplotype of the maker under evaluation, \(\hat{\gamma}_i\) is a vector of the allele effects of the \(i^{th}\) marker of \(f\) subpopulations, \(\xi\) is a vector of length \(n\) corresponding to the polygenic term and \(\epsilon\) is a vector of residuals with length \(n\).</p>

<h2>Meta-analysis</h2>

<p>For each \(i\) marker, the meta-analysis is based upon the concept of sufficient statistics, assuming that environments are independent and all information in each environment can be expressed by the allele effects \(\hat{\gamma}_i\) and the observed residual matrix \(R_i\) obtained from the association analysis. The meta-analysis attempt to verify whether the genetic (\(G\)) and environmental (\(E\)) components of \(\gamma\) differ from zero and, in addition, to verify the existance of \(G\times E\) component. In this step, the set of \(\hat{\gamma}\) from the association analyses becomes the response variable, a vector with length \(e\times f\). For the \(i^{th}\) marker, variance components are obtained from the following random model: \(\gamma_i = \mu_i + Z\alpha_i + W\beta_i + H\delta_i + e_i\). Where \(\mu_i\) is the intercept, \(Z\) is a \(ef\times f\) incidence matrix indicating the allele source, \(\alpha_i\) is the genetic effect associated to the marker, \(W\) is a \(ef\times e\) incidence matrix indicating the environmental factor, \(\beta_i\) is the coefficient associated to each environment, \(H\) is the incidence matrix of genotype by environment interaction, \(\delta_i\) is the coefficient associated to the \(G\times E\) term, and \(e_i\) is the vector of residuals with a known residual covariance matrix \(R\), a block diagonal matrix \(ef\times ef\).</p>

<h2>AMMI term</h2>

<p>The \(G\times E\) term might saturate the model once each regression coefficient \(\gamma\) is observed as an unreplicated combination of genotype and environment. The saturation does not occur because the residuals are not independent and the structure is known. Yet, there exist an alternative reparameterization of this term: the additive main effect and multiplicative interaction (AMMI) term. The AMMI term works as follows: Suppose that the analysis are being performed in a dataset with \(f=5\) subpopulations and \(e=4\) environments. Once \(\gamma\) has been estimated from the association analysis (step 1) and variance components of the genetic and environmental have been estimated with meta-analysis (step 2), then one can build the following \(E\) matrix of residuals that also contains the higher-order interaction term:</p>

<table><thead>
<tr>
<th align="left"></th>
<th align="left">E1</th>
<th align="left">E2</th>
<th align="left">E3</th>
<th align="left">E4</th>
</tr>
</thead><tbody>
<tr>
<td align="left">G1</td>
<td align="left">\(\varepsilon_{11}\)</td>
<td align="left">\(\varepsilon_{21}\)</td>
<td align="left">\(\varepsilon_{31}\)</td>
<td align="left">\(\varepsilon_{41}\)</td>
</tr>
<tr>
<td align="left">G2</td>
<td align="left">\(\varepsilon_{12}\)</td>
<td align="left">\(\varepsilon_{22}\)</td>
<td align="left">\(\varepsilon_{32}\)</td>
<td align="left">\(\varepsilon_{42}\)</td>
</tr>
<tr>
<td align="left">G3</td>
<td align="left">\(\varepsilon_{13}\)</td>
<td align="left">\(\varepsilon_{23}\)</td>
<td align="left">\(\varepsilon_{33}\)</td>
<td align="left">\(\varepsilon_{43}\)</td>
</tr>
<tr>
<td align="left">G4</td>
<td align="left">\(\varepsilon_{14}\)</td>
<td align="left">\(\varepsilon_{24}\)</td>
<td align="left">\(\varepsilon_{34}\)</td>
<td align="left">\(\varepsilon_{44}\)</td>
</tr>
<tr>
<td align="left">G5</td>
<td align="left">\(\varepsilon_{15}\)</td>
<td align="left">\(\varepsilon_{25}\)</td>
<td align="left">\(\varepsilon_{35}\)</td>
<td align="left">\(\varepsilon_{45}\)</td>
</tr>
</tbody></table>

<p>The AMMI term is extracted from the singlar-value decomposition (SVD). The SVD procedure is commonly used for the extraction of signals from non-square matrices. The decomposition is \(E = UDS'\). Where, \(U\) is a \(e\times e\) matrix, \(D\) is a \(e\times f\) retangular diagonal matrix, and \(S\) is a \(f\times f\) matrix. In analogy to the Eigendecomposition, \(U\) and \(S\) represent Eigenvectors while \(D\) are Eigenvalues. Likewise, a small fraction of principal components contain the most amount of information to recontruct the original matrix. Suppose one recontructs \(E\) using the first \(p=2\) principal components:</p>

<table><thead>
<tr>
<th align="left"></th>
<th align="left">E1</th>
<th align="left">E2</th>
<th align="left">E3</th>
<th align="left">E4</th>
</tr>
</thead><tbody>
<tr>
<td align="left">G1</td>
<td align="left">\(q_{11}\)</td>
<td align="left">\(q_{21}\)</td>
<td align="left">\(q_{31}\)</td>
<td align="left">\(q_{41}\)</td>
</tr>
<tr>
<td align="left">G2</td>
<td align="left">\(q_{12}\)</td>
<td align="left">\(q_{22}\)</td>
<td align="left">\(q_{32}\)</td>
<td align="left">\(q_{42}\)</td>
</tr>
<tr>
<td align="left">G3</td>
<td align="left">\(q_{13}\)</td>
<td align="left">\(q_{23}\)</td>
<td align="left">\(q_{33}\)</td>
<td align="left">\(q_{43}\)</td>
</tr>
<tr>
<td align="left">G4</td>
<td align="left">\(q_{14}\)</td>
<td align="left">\(q_{24}\)</td>
<td align="left">\(q_{34}\)</td>
<td align="left">\(q_{44}\)</td>
</tr>
<tr>
<td align="left">G5</td>
<td align="left">\(q_{15}\)</td>
<td align="left">\(q_{25}\)</td>
<td align="left">\(q_{35}\)</td>
<td align="left">\(q_{45}\)</td>
</tr>
</tbody></table>

<p>The matrix above can be rearranged as a vector, and be included into the model of meta-analysis replacing the current \(G\times E\) term. Thus, the model for the \(i^{th}\) marker can be also expressed as \(\gamma_i = \mu_i + Z\alpha_i + W\beta_i + Q\tau_i + e_i\).</p>

<h2>Hypothesis testing</h2>

<p>The log-likelihood of the model is, therefore, \(L(\mu,\sigma^2_\alpha,\sigma^2_\beta,\sigma^2_\tau) = -0.5( log|V|+(y-\mu)^TV^{-1}(y-\mu) )\), where the variance is expressed as \(\sigma^2_\gamma = V = ZZ^T\sigma^2_\alpha + WW^T\sigma^2_\beta + QQ^T\sigma^2_\tau + R\) and the log-likelihood of the model above is tested against \(L(\mu=\sigma^2_\alpha=\sigma^2_\beta=\sigma^2_\tau=0)\), providing the evidence that at least one of the coefficients (intercept and variance components) is not null. Thus, \(LRT = -2(L_{\mu,\sigma^2_\alpha,\sigma^2_\beta,\sigma^2_\tau}-L_{0,0,0,0})\).</p>

<h2>Woodbury&#39;s matrix identities</h2>

<p>The computational burden associated to the analysis above is originated from the determinant and inversion of the covariance matrix \(V\), a square matrix with \(ef\) rows and columns. Let \(X=[Z\sigma_\alpha || W\sigma_\beta || Q\sigma_\tau]\), such that \(V = XX^T+R\). Using the Woodbury&#39;s matrix identities, we have \(V^{-1} = R^{-1}-R^{-1}X(X^TR^{-1}X+I)^{-1}X^TR^{-1}\) and \(|V| = |X^TR^{-1}X+I||R|\), where the square matrix to be inverted has dimension \(e+f+p\). For the example, in the analysis of a dataset with \(e=18\) environments, \(f=41\) subpopulations and using \(p=2\) principal components for the \(G\times E\) term, we invert a square matrix with dimension \(18+41+2=61\) rows and columns intead of a matrix with \(18\times 41=738\) rows and columns.</p>

</body>

</html>
