<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Anna Hutchinson" />


<title>From GWAS Summary Statistics to Credible Sets</title>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">From GWAS Summary Statistics to Credible Sets</h1>
<h4 class="author">Anna Hutchinson</h4>



<div id="z-scores-to-pps" class="section level3">
<h3>Z scores to PPs</h3>
<p>Maller et al. derive a method to calculate PPs from GWAS summary statistics (<a href="https://www.ncbi.nlm.nih.gov/pubmed/23104008">Supplementary text</a>) from which the following is based on. Let <span class="math inline">\(\beta_i\)</span> for <span class="math inline">\(i=1,...,k\)</span> SNPs in a genomic region, be the regression coefficient from a single-SNP logistic regression model, quantifying the evidence of an association between SNP <span class="math inline">\(i\)</span> and the disease. Assuming that there is only one CV per region and that this is typed in the study, then if SNP <span class="math inline">\(i\)</span> is causal, <span class="math inline">\(\beta_i\neq 0\)</span> and <span class="math inline">\(\beta_j\)</span> (for <span class="math inline">\(j\neq i\)</span>) is non-zero only through LD between SNPs <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>. Note that no parametric assumptions are required for <span class="math inline">\(\beta_i\)</span> yet, so we write that it is sampled from some distribution, <span class="math inline">\(\beta_i \sim \text{[ ]}\)</span>. The likelihood is then, <span class="math display">\[\begin{equation}
\begin{split}
        P(D|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) &amp; = P(D_i
|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) \times P(D_{-i}|D_i,\text{ }\beta_i\sim\text{[ ]},\text{ }i\text{ causal})\\
    &amp; = P(D_i
|\beta_i\sim\text{[ ]},\text{ }i\text{ causal}) \times P(D_{-i}|D_i,\text{ }i\text{ causal})\,,
\end{split}
    \end{equation}\]</span></p>
<p>since <span class="math inline">\(D_{-i}\)</span> is independent of <span class="math inline">\(\beta_i\)</span> given <span class="math inline">\(D_i\)</span>. Here, <span class="math inline">\(D\)</span> is the genotype data (0, 1 or 2 counts of the minor allele) for the entire genomic region and <span class="math inline">\(i\)</span> is a SNP in the region, such that <span class="math inline">\(D_i\)</span> and <span class="math inline">\(D_{-i}\)</span> are the genotype data at SNP <span class="math inline">\(i\)</span> and at the remaining SNPs in the genomic region, respectively.</p>
<p>Parametric assumptions can now be placed on SNP <span class="math inline">\(i\)</span>’s true effect on disease. This is typically quantified as log odds ratio, and is assumed to be sampled from a Gaussian distribution, <span class="math inline">\(\beta_i\sim N(0,W)\)</span>, where <span class="math inline">\(W\)</span> is chosen to reflect the researcher’s prior belief on the variability of the true OR. Conventionally <span class="math inline">\(W\)</span> is set to <span class="math inline">\(0.2^2\)</span>, reflecting a belief that 95% of odds ratios range from <span class="math inline">\(exp(-1.96\times 0.2)=0.68\)</span> to <span class="math inline">\(exp(1.96\times 0.2)=1.48\)</span>.</p>
<p>The posterior probabilities of causality for each SNP <span class="math inline">\(i\)</span> in an associated genomic region with <span class="math inline">\(k\)</span> SNPs can be calculated where, <span class="math display">\[\begin{equation}
        PP_i=P(\beta_i \sim N(0,W),\text{ }i \text{ causal}|D)\,, \quad i \in \{1,...,k\}.
\end{equation}\]</span></p>
<p>Under the assumption that each SNP is  equally likely to be causal, then <span class="math display">\[\begin{equation}
P(\beta_i \sim N(0,W),\text{ }i\text{ causal})=\dfrac{1}{k}\,, \quad i \in \{1,...,k\}
\end{equation}\]</span> and Bayes theorem can be used to write <span class="math display">\[\begin{equation}
\begin{aligned}
    PP_i=P(\beta_i \sim N(0,W),\text{ }i \text{ causal}|D)\propto P(D|\beta_i\sim N(0,W),\text{ }i\text{ causal}).
\end{aligned}
\end{equation}\]</span></p>
<p>Dividing through by the probability of the genotype data given the null model of no genetic effect, <span class="math inline">\(H_0\)</span>, yields a likelihood ratio, <span class="math display">\[\begin{equation}
    PP_i\propto \dfrac{P(D|\beta_i \sim N(0,W),\text{ }i \text{ causal)}}{P(D|H_0)},
\end{equation}\]</span></p>
<p>from which Equation (1) can be used to derive, <span class="math display">\[\begin{equation}
    PP_i\propto \frac{P(D_i|\beta_i \sim N(0,W),\text{ }i \text{ causal})}{P(D_i|H_0)}= BF_i\,,
\end{equation}\]</span> where <span class="math inline">\(BF_i\)</span> is the Bayes factor for SNP <span class="math inline">\(i\)</span>, measuring the ratio of the probabilities of the data at SNP <span class="math inline">\(i\)</span> given the alternative (SNP <span class="math inline">\(i\)</span> is causal) and the null (no genetic effect) models.</p>
<p>In genetic association studies where sample sizes are usually large, these BFs can be approximated using Wakefield’s asymptotic Bayes factors (ABFs). Given that <span class="math inline">\(\hat\beta_i\sim N(\beta_i,V_i)\)</span> and  <span class="math inline">\(\beta_i\sim N(0,W)\)</span>,</p>
<p><span class="math display">\[\begin{equation}
    PP_i\propto BF_i \approx ABF_i=\sqrt{\frac{V_i}{V_i+W}}exp\left(\frac{Z_i^2}{2}\frac{W}{(V_i+W)}\right)\,,
\end{equation}\]</span> where <span class="math inline">\(Z_i^2=\dfrac{\hat\beta_i^2}{V_i}\)</span> is the squared marginal <span class="math inline">\(Z\)</span> score for SNP <span class="math inline">\(i\)</span>.</p>
<p>In Bayesian fine-mapping, PPs are calculated for all SNPs in the genomic region and the variants are sorted into descending order of their PP. The PPs are then cumulatively summed until some threshold, <span class="math inline">\(\alpha\)</span>, is exceeded. The variants required to exceed this threshold form the <span class="math inline">\(\alpha\)</span>-level credible set.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
