<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Anna Hutchinson" />


<title>Corrected Credible Set</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Corrected Credible Set</h1>
<h4 class="author">Anna Hutchinson</h4>



<p>This vignette will show users how the <code>corrcoverage</code> R package can be used to obtain a new credible set of variants that contains the true causal variant with some specified desired coverage value whilst containing as few variants as possible.</p>
<hr />
<div id="simulate-gwas-summary-statistics" class="section level3">
<h3>1. Simulate GWAS summary statistics</h3>
<hr />
<p>As in the <a href="https://annahutch.github.io/corrcoverage/articles/corrected-coverage.html">corrected coverage vignette</a>, let’s begin by simulating some GWAS data using the <code>simGWAS</code> package. This package is not avaliable on CRAN so the example data generated by the simGWAS package is contained in this package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">18</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(corrcoverage)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co"># library(simGWAS)</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"></a>
<a class="sourceLine" id="cb1-5" data-line-number="5">data &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="dt">package=</span><span class="st">&quot;corrcoverage&quot;</span>)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#  Simulate reference haplotypes</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">nsnps &lt;-<span class="st"> </span><span class="dv">200</span></a>
<a class="sourceLine" id="cb1-9" data-line-number="9">nhaps &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb1-10" data-line-number="10">lag &lt;-<span class="st"> </span><span class="dv">30</span>  <span class="co"># genotypes are correlated between neighbouring variants</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11">maf &lt;-<span class="st"> </span><span class="kw">runif</span>(nsnps <span class="op">+</span><span class="st"> </span>lag, <span class="fl">0.05</span>, <span class="fl">0.5</span>)  <span class="co"># common SNPs</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12">laghaps &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&quot;cbind&quot;</span>, <span class="kw">lapply</span>(maf, <span class="cf">function</span>(f) <span class="kw">rbinom</span>(nhaps, <span class="dv">1</span>, f)))</a>
<a class="sourceLine" id="cb1-13" data-line-number="13">haps &lt;-<span class="st"> </span>laghaps[, <span class="dv">1</span><span class="op">:</span>nsnps]</a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>lag) haps &lt;-<span class="st"> </span>haps <span class="op">+</span><span class="st"> </span>laghaps[, (<span class="dv">1</span><span class="op">:</span>nsnps) <span class="op">+</span><span class="st"> </span>j]</a>
<a class="sourceLine" id="cb1-15" data-line-number="15">haps &lt;-<span class="st"> </span><span class="kw">round</span>(haps<span class="op">/</span><span class="kw">matrix</span>(<span class="kw">apply</span>(haps, <span class="dv">2</span>, max), nhaps, nsnps, <span class="dt">byrow =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-16" data-line-number="16">snps &lt;-<span class="st"> </span><span class="kw">colnames</span>(haps) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;s&quot;</span>, <span class="dv">1</span><span class="op">:</span>nsnps)</a>
<a class="sourceLine" id="cb1-17" data-line-number="17">freq &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(haps <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-18" data-line-number="18">freq<span class="op">$</span>Probability &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw">nrow</span>(freq)</a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="kw">sum</span>(freq<span class="op">$</span>Probability)</a></code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">MAF &lt;-<span class="st"> </span><span class="kw">colMeans</span>(freq[, snps] <span class="op">-</span><span class="st"> </span><span class="dv">1</span>)  <span class="co"># minor allele frequencies</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">CV &lt;-<span class="st"> </span><span class="kw">sample</span>(snps[<span class="kw">which</span>(<span class="kw">colMeans</span>(haps) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.1</span>)], <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">iCV &lt;-<span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;s&quot;</span>, <span class="st">&quot;&quot;</span>, CV)  <span class="co"># index of cv</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4">LD &lt;-<span class="st"> </span><span class="kw">cor2</span>(haps) <span class="co"># correlation between SNPs</span></a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">OR &lt;-<span class="st"> </span><span class="fl">1.1</span> <span class="co"># odds ratios</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">N0 &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="co"># number of controls</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">N1 &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="co"># number of cases</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  </a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#z0 &lt;- simulated_z_score(N0 = N0, # number of controls</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#                        N1 = N1, # number of cases</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#                        snps = snps, # column names in freq</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#                        W = CV, # causal variants, subset of snps</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#                        gamma.W = log(OR), # log odds ratios</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#                        freq = freq) # reference haplotypes</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"></a>
<a class="sourceLine" id="cb4-12" data-line-number="12">z0 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">paste0</span>(data,<span class="st">&quot;/z0_2.RDS&quot;</span>))</a></code></pre></div>
<p>To calculate <span class="math inline">\(V\)</span>, the prior variance for the estimated effect size, we use <code>Var.data.cc</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">varbeta &lt;-<span class="st"> </span><span class="kw">Var.data.cc</span>(<span class="dt">f =</span> MAF, <span class="dt">N =</span> N1<span class="op">+</span>N0, <span class="dt">s =</span> N1<span class="op">/</span>(N0<span class="op">+</span>N1)) <span class="co"># variance of </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">                                                       <span class="co"># estimated effect size</span></a></code></pre></div>
<p>We can then use the <code>ppfunc</code> function to calculate the posterior probabilities of causality for each variant.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">postprobs &lt;-<span class="st"> </span><span class="kw">ppfunc</span>(<span class="dt">z =</span> z0, <span class="dt">V =</span> varbeta) </a></code></pre></div>
<p>We use the <code>est_mu</code> function to obtain an estimate of the true effect at the causal variant.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">muhat &lt;-<span class="st"> </span><span class="kw">est_mu</span>(z0, MAF, N0, N1)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">muhat</a></code></pre></div>
<pre><code>## [1] 4.970273</code></pre>
<hr />
</div>
<div id="derive-corrected-coverage-estimate" class="section level3">
<h3>2. Derive corrected coverage estimate</h3>
<hr />
<p>The <code>corrected_cov</code> function is used to find the corrected coverage of a credible set with specified threshold, say 0.9.</p>
<p>Note that this function is similar to using <code>corrcov</code> as explained in the <a href="https://annahutch.github.io/corrcoverage/articles/corrected-coverage.html">“Corrected Coverage” vignette</a>; which would require <span class="math inline">\(Z\)</span>-scores, minor allele frequencies and sample sizes. Here, we have already calculated some of the intermediaries calculated in the <code>corrcov</code> function (muhat, varbeta etc.) so we can use <code>corrected_cov</code> instead.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">thr =<span class="st"> </span><span class="fl">0.9</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">corrcov &lt;-<span class="st"> </span><span class="kw">corrected_cov</span>(<span class="dt">pp0 =</span> postprobs, <span class="dt">mu =</span> muhat, <span class="dt">V =</span> varbeta, </a>
<a class="sourceLine" id="cb9-3" data-line-number="3">                         <span class="dt">Sigma =</span> LD, <span class="dt">thr =</span> thr, <span class="dt">nrep =</span> <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">cs &lt;-<span class="st"> </span><span class="kw">credset</span>(<span class="dt">pp =</span> postprobs, <span class="dt">thr =</span> thr)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="kw">data.frame</span>(<span class="dt">claimed.cov =</span> cs<span class="op">$</span>claimed.cov, <span class="dt">corr.cov =</span>  corrcov, <span class="dt">nvar =</span> cs<span class="op">$</span>nvar)</a></code></pre></div>
<pre><code>##   claimed.cov corr.cov nvar
## 1   0.9307395 0.967282    9</code></pre>
<p>Using the Bayesian approach for statistical fine-mapping we obtain a 90% credible set consisting of 9 variants. The claimed coverage of this credible set is ~0.93, yet the corrected coverage estimate is ~0.97, suggesting that we can afford to be ‘more confident’ that we have captured the causal variant in our credible set.</p>
<hr />
</div>
<div id="evaluate-accuracy-of-estimate" class="section level3">
<h3>3. Evaluate accuracy of estimate</h3>
<hr />
<p>Again, for the purpose of this vignette we can investigate how accurate this estimate is by simulating many credible sets from the same system, and finding the proportion of these that contain the true causal variant. Note that nrep is relatively low in this example to keep stored data in the R package within CRAN guidelines. Please increase nrep to increase accuracy of estimated empirical coverage.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># z0.tmp &lt;- simulated_z_score(N0 = N0, # number of controls</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="co">#                            N1 = N1, # number of cases</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="co">#                            snps = snps, # column names in freq</span></a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="co">#                            W = CV, # causal variants, subset of snps</span></a>
<a class="sourceLine" id="cb11-5" data-line-number="5"><span class="co">#                            gamma.W = log(OR), # log odds ratios</span></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="co">#                            freq = freq, # reference haplotypes</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="co">#                            nrep = 200) # increase nrep to increase accuracy of estimate</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"></a>
<a class="sourceLine" id="cb11-9" data-line-number="9">z0.tmp &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">paste0</span>(data,<span class="st">&quot;/z0.tmp_2.RDS&quot;</span>))</a>
<a class="sourceLine" id="cb11-10" data-line-number="10"></a>
<a class="sourceLine" id="cb11-11" data-line-number="11">pps &lt;-<span class="st"> </span><span class="kw">ppfunc.mat</span>(<span class="dt">zstar =</span> z0.tmp, <span class="dt">V =</span> varbeta)  <span class="co"># find pps </span></a>
<a class="sourceLine" id="cb11-12" data-line-number="12">cs.cov &lt;-<span class="st"> </span><span class="kw">apply</span>(pps, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">credset</span>(x, <span class="dt">CV =</span> iCV, <span class="dt">thr =</span> thr)<span class="op">$</span>cov)</a>
<a class="sourceLine" id="cb11-13" data-line-number="13">true.cov.est &lt;-<span class="st"> </span><span class="kw">mean</span>(cs.cov)</a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="kw">data.frame</span>(<span class="dt">claimed.cov =</span> cs<span class="op">$</span>claimed.cov, <span class="dt">corr.cov =</span>  corrcov, </a>
<a class="sourceLine" id="cb11-15" data-line-number="15">           <span class="dt">true.cov =</span> true.cov.est, <span class="dt">nvar =</span> cs<span class="op">$</span>nvar)</a></code></pre></div>
<pre><code>##   claimed.cov corr.cov true.cov nvar
## 1   0.9307395 0.967282     0.97    9</code></pre>
<p>We find that our corrected coverage value is very close to the empirical coverage of the credible set.</p>
<hr />
</div>
<div id="obtain-a-corrected-credible-set" class="section level3">
<h3>4. Obtain a corrected credible set</h3>
<hr />
<p>Our results suggest that we may be able to remove some variants from the credible set, whilst still achieving the desired coverage of 90%.</p>
<p>The <code>corrected_cs</code> function uses GWAS summary statistics and some user-defined parameters to find the smallest credible set such that the coverage estimate is within some accuracy of the desired coverage.</p>
<p>The function requires the following parameters to be specified:</p>
<ul>
<li><code>z</code> (vector of marginal <span class="math inline">\(Z\)</span>-scores)</li>
<li><code>f</code> (vector of minor allele frequencies)</li>
<li><code>N0</code>, <code>N1</code> (number of controls and cases respectively)</li>
<li><code>Sigma</code> (SNP correlation matrix)</li>
<li><code>desired.cov</code> (desired coverage of the credible set)</li>
</ul>
<p>In addition, there are a number of optional parameters:</p>
<ul>
<li><code>lower</code> (the lower value to try for the threshold)</li>
<li><code>upper</code> (the upper value to try for the threshold)</li>
<li><code>acc</code> (the accuracy required for the coverage of the resultant credible set)</li>
<li><code>max.iter</code> (maximum number of iterations)</li>
</ul>
<p>The function uses the <a href="https://en.wikipedia.org/wiki/Root-finding_algorithm#Bisection_method">bisection root finding method</a> to converge to the smallest threshold such that the corrected coverage is larger than the desired coverage. The <code>lower</code> and <code>upper</code> parameters define the boundaries of the threshold values for the root finding method, with default values of 0 and 1 respectively. The <code>acc</code> parameter has a default value set to 0.005, meaning that the algorithm will keep attempting to find a corrected credible set (until the number of iterations reaches <code>max.iter</code>) that has coverage within 0.005 of the desired coverage value (e.g. between 0.895 and 0.905 for a 90% credible set).</p>
<p>The function reports the threshold values tested and their corresponding corrected coverage value at each iteration. The maximum number of iterations for the bisecting root finding algorithm is an optional parameter, with default value 20. The functions stops when either the number of iterations reaches the maximum, or the corrected coverage is within some accuracy of the desired coverage.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">res &lt;-<span class="st"> </span><span class="kw">corrected_cs</span>(<span class="dt">z =</span> z0, <span class="dt">f =</span> MAF, N0, N1, </a>
<a class="sourceLine" id="cb13-2" data-line-number="2">                    <span class="dt">Sigma =</span> LD, <span class="dt">lower =</span> <span class="fl">0.5</span>, <span class="dt">upper =</span> <span class="dv">1</span>, <span class="dt">desired.cov =</span> <span class="fl">0.9</span>)</a></code></pre></div>
<pre><code>## [1] &quot;thr:  0.75 , cov:  0.918448964618699&quot;
## [1] &quot;thr:  0.625 , cov:  0.870505410370508&quot;
## [1] &quot;thr:  0.6875 , cov:  0.893551309455069&quot;
## [1] &quot;thr:  0.71875 , cov:  0.906172863191769&quot;
## [1] &quot;thr:  0.703125 , cov:  0.899644208037482&quot;
## [1] &quot;thr:  0.7109375 , cov:  0.901677186379133&quot;</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">res</a></code></pre></div>
<pre><code>## $credset
## [1] &quot;s54&quot; &quot;s58&quot; &quot;s57&quot; &quot;s67&quot;
## 
## $req.thr
## [1] 0.7109375
## 
## $corr.cov
## [1] 0.9016772
## 
## $size
## [1] 0.7343958</code></pre>
<p>The first threshold value tested is the midpoint of the <code>lower</code> (0.5) and <code>upper</code> (1) parameter values, which here is 0.75. The coverage of this credible set is too high (<span class="math inline">\(&gt;0.9\)</span>) and so a smaller threshold value is tested, the midpoint of the <code>lower</code> parameter value (0.5) and the previously tried threshold value (0.75). This credible set has a coverage that is too low, and so a higher threshold value is tested, and so on.</p>
<p>In this example we see that a much smaller threshold value is required to obtain a credible set with 90% corrected coverage of the causal variant, containing only 4 variants. In the standard Bayesian approach, the threshold value of 90% leads to over-coverage.</p>
<hr />
<p>Finally, we can compare this coverage estimate of the corrected credible set to an empirical estimate of the true coverage.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">new.cs.sims &lt;-<span class="st"> </span><span class="kw">apply</span>(pps, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">credset</span>(x, <span class="dt">CV =</span> iCV, <span class="dt">thr =</span> res<span class="op">$</span>req.thr)<span class="op">$</span>cov)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">true.cov.est2 &lt;-<span class="st"> </span><span class="kw">mean</span>(new.cs.sims)</a></code></pre></div>
<p>Original 90% credible set:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">df1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">claimed.cov =</span> <span class="kw">round</span>(cs<span class="op">$</span>claimed.cov, <span class="dv">3</span>), <span class="dt">corr.cov =</span>  <span class="kw">round</span>(corrcov, <span class="dv">3</span>), <span class="dt">true.cov =</span> <span class="kw">round</span>(true.cov.est, <span class="dv">3</span>), <span class="dt">nvar =</span> cs<span class="op">$</span>nvar)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw">print</span>(df1, <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>##  claimed.cov corr.cov true.cov nvar
##        0.931    0.967     0.97    9</code></pre>
<p>New 90% credible set:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">df2 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">claimed.cov =</span> <span class="kw">round</span>(res<span class="op">$</span>size, <span class="dv">3</span>), <span class="dt">corr.cov =</span> <span class="kw">round</span>(res<span class="op">$</span>corr.cov, <span class="dv">3</span>), <span class="dt">true.cov =</span> <span class="kw">round</span>(true.cov.est2, <span class="dv">3</span>), <span class="dt">nvar =</span> <span class="kw">length</span>(res<span class="op">$</span>credset))</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">print</span>(df2, <span class="dt">row.names =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>##  claimed.cov corr.cov true.cov nvar
##        0.734    0.902      0.9    4</code></pre>
<hr />
<p>This vignette has shown how the <code>corrcoverage</code> R package can be used to improve the resolution of a credible set from Bayesian genetic fine-mapping, without the use of any additional data.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
